proposals/patterns3.md

```diff
@@ -0,0 +1,253 @@
+
+When the input is a type for which a suitable built-in binary relational operator is defined that is applicable with the input as its left operand and the given constant as its right operand, the evaluation of that operator is taken as the meaning of the relational pattern.  Otherwise we convert the input to the type of the expression using an explicit nullable or unboxing conversion.  It is a compile-time error if no such conversion exists.  The pattern is considered not to match if the conversion fails.  If the conversion succeeds then the result of the pattern-matching operation is the result of evaluating the expression `e OP v` where `e` is the converted input, `OP` is the relational operator, and `v` is the constant expression.
+
+## Pattern Combinators
+
+Pattern *combinators* permit matching both of two different patterns using `and` (this can be extended to any number of patterns by the repeated use of `and`), either of two different patterns using `or` (ditto), or the *negation* of a pattern using `not`.
+
+A common use of a combinator will be the idiom
+
+``` c#
+if (e is not null) ...
+```
+
+More readable than the current idiom `e is object`, this pattern clearly expresses that one is checking for a non-null value.
+
+The `and` and `or` combinators will be useful for testing ranges of values
+
+``` c#
+bool IsLetter(char c) => c is >= 'a' and <= 'z' or >= 'A' and <= 'Z';
+```
+
+This example illustrates our expectation that `and` will have a higher parsing priority (i.e. will bind more closely) than `or`.  The programmer can use the *parenthesized pattern* to make the precedence explicit:
+
+``` c#
+bool IsLetter(char c) => c is (>= 'a' and <= 'z') or (>= 'A' and <= 'Z');
+```
+
+Like all patterns, these combinators can be used in any context in which a pattern is expected, including nested patterns, the *is-pattern-expression*, the *switch-expression*, and the pattern of a switch statement's case label.
+
+```antlr
+pattern
+    : disjunctive_pattern
+    ;
+disjunctive_pattern
+    : disjunctive_pattern 'or' conjunctive_pattern
+    | conjunctive_pattern
+    ;
+conjunctive_pattern
+    : conjunctive_pattern 'and' negated_pattern
+    | negated_pattern
+    ;
+negated_pattern
+    : 'not' negated_pattern
+    | primary_pattern
+    ;
+primary_pattern
+    : // all of the patterns forms previously defined
+    ;
+```
+
+## Open Issues with Proposed Changes
+
+### Syntax for relational operators
+
+Are `and`, `or`, and `not` some kind of contextual keyword?  If so, is there a breaking change (e.g. compared to their use as a designator in a *declaration-pattern*).
+
+Should we support some combination of declaration pattern along with a relational pattern?  For example,
+
+``` csharp
+if (o is int x <= 100) // x is an int with value < 100 here.
```

Comment:
fwiw, looking at this code sample I really don't like it. I much prefer the `and` version.

Reply:
Me too, which is why we didn't do this.